### PR TITLE
updated blob.bypass_serialization patch

### PR DESCRIPTION
**Pull request by ixcat**
per feeedback in #578

### CODE REVIEW

**Comment by ixcat** [comment]
one remaining task is failure mode - not sure best way / level to do this -

Could just sanity check the pack/unpack data checks..

Manually injecting bad data into a table seems like alot of work, but could be done if deemed necessary.

File: datajoint/blob.py [hunk 1]
```diff
@@ -210,6 +213,11 @@ def __str__(self):
 
 
 def pack(obj, compress=True):
+    if bypass_serialization is True:
```
Comment (dimitri-yatsenko) [inline datajoint/blob.py]:
```suggestion
    if bypass_serialization:
```

File: datajoint/blob.py [hunk 2]
```diff
@@ -210,6 +213,11 @@ def __str__(self):
...
+        assert(any([obj[:len(pattern)] == pattern
```
Comment (dimitri-yatsenko) [inline datajoint/blob.py]:
```suggestion
        assert obj.startswith(tuple(decode_loookup))
```
`assert` is not a function. Any can take an iterator and quit as soon as condition is met without forming a list. `str.startswith` can take a tuple of strings.

File: datajoint/blob.py [hunk 3]
```diff
@@ -210,6 +213,11 @@ def __str__(self):
...
+               for pattern in decode_lookup]))
```
Comment (dimitri-yatsenko) [inline datajoint/blob.py]:
```suggestion
```

File: datajoint/blob.py [hunk 4]
```diff
@@ -300,8 +308,13 @@ def pack_dict(obj):
 
 
 def unpack(blob, **kwargs):
+
     if blob is None:
         return None
 
-    return BlobReader(blob, **kwargs).unpack()
+    if bypass_serialization is True:
+        assert(any([blob[:len(pattern)] == pattern
```
Comment (dimitri-yatsenko) [inline datajoint/blob.py]:
```suggestion
        assert blob.startswith(tuple(decode_lookup))
```

File: datajoint/blob.py [hunk 5]
```diff
@@ -300,8 +308,13 @@ def pack_dict(obj):
...
+               for pattern in decode_lookup]))
```
Comment (dimitri-yatsenko) [inline datajoint/blob.py]:
```suggestion
```

File: tests/test_bypass_serialization.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import numpy as np
+
+schema_in = dj.schema(PREFIX + '_test_bypass_serialization_in',
+                      connection=dj.conn(**CONN_INFO))
+
+schema_out = dj.schema(PREFIX + '_test_blob_bypass_serialization_out',
+                       connection=dj.conn(**CONN_INFO))
+
+
+tst_blob = np.array([1, 2, 3])  # test blob;
```
Comment (dimitri-yatsenko) [inline tests/test_bypass_serialization.py]:
```suggestion
test_blob = np.array([1, 2, 3]) 
```

File: tests/test_bypass_serialization.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+def test_bypass_serialization():
+    dj.blob.bypass_serialization = True
+    OutputTable.insert(InputTable.fetch(as_dict=True))
+    dj.blob.bypass_serialization = True
+
+    ins = InputTable.fetch(as_dict=True)
+    outs = OutputTable.fetch(as_dict=True)
+
+    assert_true(all(np.array_equals(i[0]['data'], tst_blob),
```
Comment (dimitri-yatsenko) [inline tests/test_bypass_serialization.py]: This `assert_true` will never fail. It is not testing what you like it to test.

Comment (dimitri-yatsenko) [inline tests/test_bypass_serialization.py]: This assumes that `ins` and `outs` are sorted the same way.  What guarantees that?

Comment (ixcat) [reply tests/test_bypass_serialization.py]: there's only one record :) but, good point. will switch to fetch1.

**Comment by dimitri-yatsenko** [comment]
progress?

**Comment by ixcat** [comment]
updated/incorporated suggestions

File: tests/test_bypass_serialization.py [hunk 8]
```diff
@@ -36,12 +36,9 @@ class OutputTable(dj.Manual):
...
-    ins = InputTable.fetch(as_dict=True)
-    outs = OutputTable.fetch(as_dict=True)
+    dj.blob.bypass_serialization = False
 
-    assert_true(all(np.array_equals(i[0]['data'], tst_blob),
-                np.array_equals(i[0]['data'], i[1]['data']))
-                for i in zip(ins, outs))
+    i = InputTable.fetch1()
+    o = OutputTable.fetch1()
+    assert_true(np.array_equal(i['data'], o['data']))
```
Comment (dimitri-yatsenko) [inline tests/test_bypass_serialization.py]: you can do assert_array_equal

File: tests/test_bypass_serialization.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    definition = """
+    id:                 int
+    ---
+    data:               blob
+    """
+
+
+def test_bypass_serialization():
+    dj.blob.bypass_serialization = True
+    OutputTable.insert(InputTable.fetch(as_dict=True))
```
Comment (dimitri-yatsenko) [inline tests/test_bypass_serialization.py]: hm.. this didn't really test the bypass

Comment (dimitri-yatsenko) [reply tests/test_bypass_serialization.py]: also, this would pass if `InputTable` were empty, no matter what.

Comment (dimitri-yatsenko) [reply tests/test_bypass_serialization.py]: ok, i will merge and add one more assert to fix this.

**Pull request merged**